### PR TITLE
Yf use mapped file dictionary v2

### PR DESCRIPTION
**Pull request by yfarjoun**
Changing MergeBamAlignment so that it merges the dictionaries of the aligned bam and the reference instead of taking the reference dict file.
This is since bwa-mem (latests versions) add a AH:* to the sequence records of "Alternate Haplotypes" and this is lost in translation otherwise.

-added a tests that reads from a named stream /dev/fd/<int> for either mapped or unmapped sam

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by yfarjoun** [comment]
@tfenne I re-did my previous PR. it now doesn't read the file twice, so it works on streams. to prove it, I added two streaming tests!

I also split the commit into two for your viewing convenience 
1. java 8 stuff, lines, braces etc.
2. substance

I hope you'll be able to give this a view. Sorry for having asked for this PR having dragged on for so long.

**Comment by yfarjoun** [comment]
sorry about that @tfenne when I did `git reset` to split the commit into two, I dropped some of the files and didn't add them back....fixed this now.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 1]
```diff
@@ -183,6 +157,7 @@ public boolean isPopulatePaTag() {
             return populatePATag;
         }
     }
+    protected abstract SAMSequenceDictionary getDictionaryForMergedBam();
```
Comment (tfenne) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Should have a newline before method declaration.

File: src/test/java/picard/sam/MergeBamAlignmentTest.java [hunk 2]
```diff
@@ -954,7 +952,7 @@ public void testEarliestFragmentStrategyPaired() throws Exception {
         alignedSam.deleteOnExit();
 
         // Populate the header with SAMSequenceRecords
-        header.getSequenceDictionary().addSequence(new SAMSequenceRecord("chr1", 1000000));
+        header.setSequenceDictionary(SamReaderFactory.makeDefault().getFileHeader(sequenceDict2).getSequenceDictionary());
```
Comment (tfenne) [inline src/test/java/picard/sam/MergeBamAlignmentTest.java]: Why not use the SAMSequenceDictionaryExtractor here and on the other similar lines?

**Comment by tfenne** [comment]
Just ran a small data set, streaming all the way from the unmapped bam, through SamToFastq, through bwa, through MergeBamAlignment built from this branch, and out to file.  All worked well!

**Comment by yfarjoun** [comment]
thanks for the extra leg work @tfenne much appreciated.

File: src/test/java/picard/sam/MergeBamAlignmentTest.java (hunk 2 above)
Comment (yfarjoun) [inline src/test/java/picard/sam/MergeBamAlignmentTest.java]: I was being lazy. thanks for the suggestion.

File: src/main/java/picard/sam/AbstractAlignmentMerger.java (hunk 1 above)
Comment (yfarjoun) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: done.

**Comment by yfarjoun** [comment]
About the merger, I am a little concerned about rocking this particular boat given the swiss-army-knife aspect that it has....how confident do we feel that all the use-cases and behaviors are actually covered by tests?

**Comment by yfarjoun** [comment]
waiting for test to pass. I'm assuming that the remaining comments were so small that after fixing them I can merge....

**Pull request merged**